tensorflow/core/kernels/sparse_tensor_dense_matmul_op_gpu.cu.cc

```diff
@@ -93,38 +130,78 @@ struct SparseTensorDenseMatMulFunctor<GPUDevice, T, Tindices, ADJ_A, ADJ_B> {
     int b_cols = b.dimension(1);
 
     const GPUDevice& d = ctx->eigen_device<GPUDevice>();
-    using Tsum = typename SumType<T>::type;
-    Tsum* maybe_temp_out_data = nullptr;
     Tensor temp_out_t;
+    using Tsum = typename SumType<T>::type;
+    using Tupcast = typename SumType<T>::upcast_type;
     bool sum_type_is_different = !std::is_same<T, Tsum>::value;
-    if (sum_type_is_different) {
-      TF_RETURN_IF_ERROR(ctx->allocate_temp(
-          DataTypeToEnum<Tsum>::value,
-          TensorShape({out.dimension(0), out.dimension(1)}), &temp_out_t));
-      auto temp_out = temp_out_t.matrix<Tsum>();
-      maybe_temp_out_data = temp_out.data();
-      temp_out.device(d) = temp_out.constant(Tsum(0));
-    } else {
-      // Note: The reinterpret cast is only required to avoid a compilation
-      // error; it is only used if Tsum == T.
-      maybe_temp_out_data = reinterpret_cast<Tsum*>(out.data());
-      out.device(d) = out.constant(T(0));
-    }
-
-    // TODO(ebrevdo): Should this be alpha * nnz instead of
-    // out.size()?  Perhaps p * nnz ?
     GpuLaunchConfig config = GetGpuLaunchConfig(p * nnz, d);
 
-    TF_CHECK_OK(GpuLaunchKernel(
-        SparseTensorDenseMatMulKernel<T, Tsum, Tindices, ADJ_A, ADJ_B>,
-        config.block_count, config.thread_per_block, 0, d.stream(), nnz, m,
-        b_rows, b_cols, p, a_indices.data(), a_values.data(), b.data(),
-        maybe_temp_out_data));
+    if (RequireDeterminism()) {
+      if (std::is_same<T, double>::value ||
+          std::is_same<T, complex128>::value) {
+        return errors::Unimplemented(
+            "No deterministic GPU implementation of "
+            "float64 or complex128 for sparse_dense_matmul");
+      }
+
+       Tupcast* maybe_temp_out_data = nullptr;
+
+       TF_RETURN_IF_ERROR(ctx->allocate_temp(
+            DataTypeToEnum<Tupcast>::value,
+            TensorShape({out.dimension(0), out.dimension(1)}), &temp_out_t));
+       auto temp_out = temp_out_t.matrix<Tupcast>();
+       maybe_temp_out_data = temp_out.data();
+
+       TF_CHECK_OK(GpuLaunchKernel(
+           SetZero<Tupcast>, config.block_count, config.thread_per_block, 0,
+           d.stream(), m * p, maybe_temp_out_data));
+
+       TF_CHECK_OK(GpuLaunchKernel(
+          SparseTensorDenseMatMulKernel<T, Tupcast, Tindices, ADJ_A, ADJ_B>,
+          config.block_count, config.thread_per_block, 0, d.stream(), nnz, m,
+          b_rows, b_cols, p, a_indices.data(), a_values.data(), b.data(),
+          maybe_temp_out_data));
```

Comment:
It might seem pedantic, but it's important to generate code in a standard way to make it easy to read and maintain. Search for, and review, all 30 instances of the word "indent" in the [Google C++ Style Guide](https://google.github.io/styleguide/cppguide.html).